third_party/move/move-binary-format/src/file_format_common.rs

```diff
@@ -524,8 +524,13 @@ pub(crate) mod versioned_data {
                 },
             };
             if version == 0 || version > u32::min(max_version, VERSION_MAX) {
-                return Err(PartialVMError::new(StatusCode::UNKNOWN_VERSION));
-            } else if version == VERSION_NEXT && !cfg!(test) && !cfg!(feature = "fuzzing") {
+                return Err(PartialVMError::new(StatusCode::UNKNOWN_VERSION)
+                    .with_message(format!("bytecode version {} unsupported", version)));
+            } else if version == VERSION_NEXT
+                && !cfg!(test)
+                && !cfg!(feature = "testing")
+                && !cfg!(feature = "fuzzing")
```

Comment:
I'd strongly recommend we move away from gating critical features by the crate feature flags, due to cargo's handling of feature unification. 
- Right now there is a single fragile test during node startup ensuring that the node binary is not built with the "testing" feature flag
- The test does not check the "fuzzing" feature flag

It would be a lot safer to pass in a bool variable indicating whether the next version shall be accepted.

Reply:
As we discussed offline, this is not feature gating. Its a double safeguard that experimental code cannot run in production. The actual feature gating of the maximally supported bytecode version continues to be done via regular feature gating and on-chain config.

Reply:
> The actual feature gating of the maximally supported bytecode version continues to be done via regular feature gating and on-chain config.

I don't see this in the current implementation though. Have I missed anything?

My worry right now is just that if someone sends a move binary with NEXT_VERSION against the current implementation, it will be accepted if the cargo features are misconfigured, which sounds dangerous for production.